### ISSUE TITLE
作者你好，我在评估指令中，以下错误怎么解决啊

### ISSUE THREAD

**Issue by hwx-418** [Junda24/MonSter#30]
(monster) root@autodl-container-c62840940c-ad7ded55:~/autodl-tmp/MonSter-main# python evaluate_stereo.py --restore_ckpt ./pretrained/eth3d.pth
xFormers not available
xFormers not available
DepthAnythingV2 out_channels [256, 512, 1024, 1024]
DepthAnythingV2_decoder out_channels [256, 512, 1024, 1024]
Total number of parameters: 388.69M
Total number of trainable parameters: 53.38M
2025-05-03 21:23:25,903 INFO     [evaluate_stereo.py:420] Loading checkpoint...
2025-05-03 21:23:25,904 INFO     [evaluate_stereo.py:421] ./pretrained/eth3d.pth
2025-05-03 21:23:26,824 INFO     [evaluate_stereo.py:436] Done loading checkpoint
The model has 53.38M learnable parameters.
/root/miniconda3/envs/monster/lib/python3.8/site-packages/numpy/core/fromnumeric.py:3464: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
/root/miniconda3/envs/monster/lib/python3.8/site-packages/numpy/core/_methods.py:192: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
Validation ETH3D: EPE nan, D1 nan

**Comment by Junda24** [comment]
> (monster) root@autodl-container-c62840940c-ad7ded55:~/autodl-tmp/MonSter-main# python evaluate_stereo.py --restore_ckpt ./pretrained/eth3d.pth xFormers not available xFormers not available DepthAnythingV2 out_channels [256, 512, 1024, 1024] DepthAnythingV2_decoder out_channels [256, 512, 1024, 1024] Total number of parameters: 388.69M Total number of trainable parameters: 53.38M 2025-05-03 21:23:25,903 INFO [evaluate_stereo.py:420] Loading checkpoint... 2025-05-03 21:23:25,904 INFO [evaluate_stereo.py:421] ./pretrained/eth3d.pth 2025-05-03 21:23:26,824 INFO [evaluate_stereo.py:436] Done loading checkpoint The model has 53.38M learnable parameters. /root/miniconda3/envs/monster/lib/python3.8/site-packages/numpy/core/fromnumeric.py:3464: RuntimeWarning: Mean of empty slice. return _methods._mean(a, axis=axis, dtype=dtype, /root/miniconda3/envs/monster/lib/python3.8/site-packages/numpy/core/_methods.py:192: RuntimeWarning: invalid value encountered in scalar divide ret = ret.dtype.type(ret / rcount) Validation ETH3D: EPE nan, D1 nan
有没有更多的报错信息，这个判断不出来是什么bug

**Comment by lingshijiang** [comment]
原本的测试有问题我是这样解决的
epe_val, out_val = accelerator.gather_for_metrics((sample_epe_mean, sample_out_mean))
                        # 安全处理收集的指标
                        if isinstance(epe_val, torch.Tensor):
                            if len(epe_val.shape) > 0:  # 张量有维度
                                elem_num += epe_val.shape[0]
                                for i in range(epe_val.shape[0]):
                                    if not torch.isnan(epe_val[i]).item() and not torch.isinf(epe_val[i]).item():
                                        total_epe += epe_val[i]
                                        total_out += out_val[i]
                            else:  # 0维张量 (标量)
                                elem_num += 1
                                if not torch.isnan(epe_val).item() and not torch.isinf(epe_val).item():
                                    total_epe += epe_val
                                    total_out += out_val
                        else:  # 非张量对象
                            elem_num += 1
                            if not np.isnan(epe_val) and not np.isinf(epe_val):
                                total_epe += epe_val
                                total_out += out_val

                        valid_samples += 1

                        # 每处理50个样本打印一次中间结果
                        if (batch_idx + 1) % 50 == 0 and elem_num > 0:
                            print(
                                f"Intermediate results after {batch_idx + 1} samples: EPE={total_epe / elem_num:.4f}, D1={100 * total_out / elem_num:.2f}%")

                    except Exception as e:
                        import traceback
                        print(f"ERROR processing validation sample {batch_idx}:")
                        print(traceback.format_exc())

                print(f"\nValidation summary:")
                print(f"  Total samples: {len(val_loader)}")
                print(f"  Successfully processed: {valid_samples} ({valid_samples / len(val_loader) * 100:.2f}%)")
                # 计算并记录最终结果
                if elem_num > 0:
                    mean_epe = total_epe / elem_num
                    mean_d1 = 100 * total_out / elem_num
                    print(f"  Final mean EPE: {mean_epe:.4f}, mean D1: {mean_d1:.2f}%")
                    accelerator.log({'val/epe': mean_epe, 'val/d1': mean_d1}, total_step)
                else:
                    print("  WARNING: No valid elements for computing validation metrics!")
                    # 记录警告而非NaN
                    accelerator.log({'val/warning': 'No valid elements for metrics'}, total_step)
                model.train()

**Comment by L190200201sunasijia** [comment]
> 原本的测试有问题我是这样解决的 epe_val, out_val = accelerator.gather_for_metrics((sample_epe_mean, sample_out_mean)) # 安全处理收集的指标 if isinstance(epe_val, torch.Tensor): if len(epe_val.shape) > 0: # 张量有维度 elem_num += epe_val.shape[0] for i in range(epe_val.shape[0]): if not torch.isnan(epe_val[i]).item() and not torch.isinf(epe_val[i]).item(): total_epe += epe_val[i] total_out += out_val[i] else: # 0维张量 (标量) elem_num += 1 if not torch.isnan(epe_val).item() and not torch.isinf(epe_val).item(): total_epe += epe_val total_out += out_val else: # 非张量对象 elem_num += 1 if not np.isnan(epe_val) and not np.isinf(epe_val): total_epe += epe_val total_out += out_val
> 
> ```
>                     valid_samples += 1
> 
>                     # 每处理50个样本打印一次中间结果
>                     if (batch_idx + 1) % 50 == 0 and elem_num > 0:
>                         print(
>                             f"Intermediate results after {batch_idx + 1} samples: EPE={total_epe / elem_num:.4f}, D1={100 * total_out / elem_num:.2f}%")
> 
>                 except Exception as e:
>                     import traceback
>                     print(f"ERROR processing validation sample {batch_idx}:")
>                     print(traceback.format_exc())
> 
>             print(f"\nValidation summary:")
>             print(f"  Total samples: {len(val_loader)}")
>             print(f"  Successfully processed: {valid_samples} ({valid_samples / len(val_loader) * 100:.2f}%)")
>             # 计算并记录最终结果
>             if elem_num > 0:
>                 mean_epe = total_epe / elem_num
>                 mean_d1 = 100 * total_out / elem_num
>                 print(f"  Final mean EPE: {mean_epe:.4f}, mean D1: {mean_d1:.2f}%")
>                 accelerator.log({'val/epe': mean_epe, 'val/d1': mean_d1}, total_step)
>             else:
>                 print("  WARNING: No valid elements for computing validation metrics!")
>                 # 记录警告而非NaN
>                 accelerator.log({'val/warning': 'No valid elements for metrics'}, total_step)
>             model.train()
> ```

你这个不是跑eval吗？怎么还有model.train()。我也报了同样的问题不知道怎么解决

**Comment by Junda24** [comment]
Issue里面有人提过相同的问题，你可以参考一下，然后那个model.train是为了重新打开训练模式，使得测试完之后继续进入训练模式来进行下一步的训练



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: Duck_in_slippers ***@***.***&gt;
发送时间: 2025年7月18日 17:58
收件人: Junda24/MonSter ***@***.***&gt;
抄送: Junda Cheng ***@***.***&gt;, Comment ***@***.***&gt;
主题: Re: [Junda24/MonSter] 作者你好，我在评估指令中，以下错误怎么解决啊 (Issue #30)
